Description: Remove staden include dir
Author: Michael Steinbaugh <mike@steinbaugh.com>
Last-Update: 2023-10-10
--- a/cmake/Modules/Findlibstadenio.cmake
+++ b/cmake/Modules/Findlibstadenio.cmake
@@ -16,15 +16,6 @@
 find_library(HTSCODEC_LIBRARY NAMES htscodecs libhtscodecs
   HINTS ${STADEN_ROOT} ENV STADEN_ROOT PATH_SUFFIXES lib lib64)
 
-if(STADEN_INCLUDE_DIR)
-  set(_version_regex "^#define[ \t]+PACKAGE_VERSION[ \t]+\"([^\"]+)\".*")
-  file(STRINGS "${STADEN_INCLUDE_DIR}/io_lib/io_lib_config.h"
-    STADEN_VERSION REGEX "${_version_regex}")
-  string(REGEX REPLACE "${_version_regex}" "\\1"
-    STADEN_VERSION "${STADEN_VERSION}")
-  unset(_version_regex)
-endif()
-
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(libstadenio DEFAULT_MSG
                                   STADEN_LIBRARY 
